skrifa/src/meta/metrics.rs

```diff
@@ -0,0 +1,434 @@
+        // spec, backward compatibility and broken fonts.
+        //
+        // We use the same strategy as FreeType:
+        // 1. Use the OS/2 metrics if the table exists and the USE_TYPO_METRICS
+        //    flag is set.
+        // 2. Otherwise, use the hhea metrics.
+        // 3. If hhea metrics are zero and the OS/2 table exists:
+        //    3a. Use the typo metrics if they are non-zero
+        //    3b. Otherwise, use the win metrics
+        let os2 = font.os2().ok();
+        let mut used_typo_metrics = false;
+        if let Some(os2) = &os2 {
+            const USE_TYPO_METRICS: u16 = 1 << 7;
+            if os2.fs_selection() & USE_TYPO_METRICS != 0 {
+                metrics.ascent = os2.s_typo_ascender() as f32 * scale;
+                metrics.descent = os2.s_typo_descender() as f32 * scale;
+                metrics.leading = os2.s_typo_line_gap() as f32 * scale;
+                used_typo_metrics = true;
+            }
+            metrics.average_width = Some(os2.x_avg_char_width() as f32 * scale);
+            metrics.cap_height = os2.s_cap_height().map(|v| v as f32 * scale);
+            metrics.x_height = os2.sx_height().map(|v| v as f32 * scale);
+            metrics.strikeout = Some(Decoration {
+                offset: os2.y_strikeout_position() as f32 * scale,
+                thickness: os2.y_strikeout_size() as f32 * scale,
+            });
+        }
+        if !used_typo_metrics {
+            if let Ok(hhea) = font.hhea() {
+                metrics.ascent = hhea.ascender().to_i16() as f32 * scale;
+                metrics.descent = hhea.descender().to_i16() as f32 * scale;
+                metrics.leading = hhea.line_gap().to_i16() as f32 * scale;
+            }
+            if metrics.ascent == 0.0 && metrics.descent == 0.0 {
+                if let Some(os2) = &os2 {
+                    if os2.s_typo_ascender() != 0 || os2.s_typo_descender() != 0 {
+                        metrics.ascent = os2.s_typo_ascender() as f32 * scale;
+                        metrics.descent = os2.s_typo_descender() as f32 * scale;
+                        metrics.leading = os2.s_typo_line_gap() as f32 * scale;
+                    } else {
+                        metrics.ascent = os2.us_win_ascent() as f32 * scale;
+                        // Win descent is always positive while other descent values are negative. Negate it
+                        // to ensure we return consistent metrics.
+                        metrics.descent = -(os2.us_win_descent() as f32 * scale);
+                    }
+                }
+            }
+        }
+        if let (Ok(mvar), true) = (font.mvar(), !coords.is_empty()) {
+            use read_fonts::tables::mvar::tags::*;
+            macro_rules! metric_delta {
+                ($tag: ident) => {
+                    mvar.metric_delta($tag, coords).unwrap_or_default().to_f64() as f32 * scale
+                };
+            }
```

Comment:
I appreciate the power macros afford but here normal code would work fine and my bias would be to prefer normal code when there isn't a good reason why not. Specifically:

```rust
// Define a free fn to lookup a delta
fn lookup_delta(mvar: &TableRef<MvarMarker>, tag: Tag, coords: &[F2Dot14], scale: f32) -> f32 {
    mvar.metric_delta(tag, coords).unwrap_or_default().to_f64() as f32 * scale
}

// OK, but I don't want to have noise from passing mvar, coords, scale!!
// So, instead of defining a macro, use a closure to capture the values
let metric_delta = |tag| lookup_delta(&mvar, tag, coords, scale);

// Now call metric_delta(HASC) instead of metric_delta!(HASC)

```

Reply:
Yep! The closure works just fine even without the intermediate function. Will change.